presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowFlightClientHandler.java

```diff
@@ -0,0 +1,127 @@
+
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.util.Optional;
+
+import static com.facebook.plugin.arrow.ArrowErrorCode.ARROW_FLIGHT_ERROR;
+
+public abstract class ArrowFlightClientHandler
+{
+    private static final Logger logger = Logger.get(ArrowFlightClientHandler.class);
+    private final ArrowFlightConfig config;
+
+    private RootAllocator allocator;
+
+    public ArrowFlightClientHandler(ArrowFlightConfig config)
+    {
+        this.config = config;
+    }
+
+    private ArrowFlightClient initializeClient(Optional<String> uri)
+    {
+        try {
+            Optional<InputStream> trustedCertificate = Optional.empty();
+
+            Location location;
+            if (uri.isPresent()) {
+                location = new Location(uri.get());
+            }
+            else {
+                if (config.getArrowFlightServerSslEnabled() != null && !config.getArrowFlightServerSslEnabled()) {
+                    location = Location.forGrpcInsecure(config.getFlightServerName(), config.getArrowFlightPort());
+                }
+                else {
+                    location = Location.forGrpcTls(config.getFlightServerName(), config.getArrowFlightPort());
+                }
+            }
+
+            if (null == allocator) {
+                initializeAllocator();
+            }
+
+            FlightClient.Builder flightClientBuilder = FlightClient.builder(allocator, location);
+            if (config.getVerifyServer() != null && !config.getVerifyServer()) {
+                flightClientBuilder.verifyServer(false);
+            }
+            else if (config.getFlightServerSSLCertificate() != null) {
+                trustedCertificate = Optional.of(new FileInputStream(config.getFlightServerSSLCertificate()));
+                flightClientBuilder.trustedCertificates(trustedCertificate.get()).useTls();
+            }
+
+            FlightClient flightClient = flightClientBuilder.build();
+            return new ArrowFlightClient(flightClient, trustedCertificate);
+        }
+        catch (Exception ex) {
+            throw new ArrowException(ARROW_FLIGHT_ERROR, "The flight client could not be obtained." + ex.getMessage(), ex);
+        }
+    }
+
+    private synchronized void initializeAllocator()
+    {
+        if (allocator == null) {
+            allocator = new RootAllocator(Long.MAX_VALUE);
+        }
+    }
+
+    protected abstract CredentialCallOption getCallOptions(ConnectorSession connectorSession);
+
+    public ArrowFlightConfig getConfig()
+    {
+        return config;
+    }
+
+    public ArrowFlightClient getClient(Optional<String> uri)
+    {
+        return initializeClient(uri);
+    }
+
+    public FlightInfo getFlightInfo(FlightDescriptor flightDescriptor, ConnectorSession connectorSession)
+    {
+        try (ArrowFlightClient client = getClient(Optional.empty())) {
+            CredentialCallOption auth = this.getCallOptions(connectorSession);
```

Comment:
```suggestion
            CredentialCallOption auth = getCallOptions(connectorSession);
```

Does this need to return an array of options?

Reply:
Updated to return an array of options.